File: modules/sql/app/model/querycorrectors/select/GroupByMatcher.java
package model.querycorrectors.select;

import java.util.function.BiPredicate;

import model.matching.Match;
import model.matching.Matcher;
import net.sf.jsqlparser.expression.Expression;
import net.sf.jsqlparser.schema.Column;

public class GroupByMatcher extends Matcher<Expression, Match<Expression>> {
  
  private static final BiPredicate<Expression, Expression> GROUP_BY_TEST = (g1, g2) -> {
    Column c1 = (Column) g1;
    Column c2 = (Column) g2;
    return c1.getColumnName().equals(c2.getColumnName());
  };
  
  public GroupByMatcher() {
    super("Group By-Elemente", GROUP_BY_TEST, GroupByMatch::new);
  }
  
}

File: app/model/tools/sql/matcher/GroupByMatcher.scala
package model.tools.sql.matcher

import model._
import model.core.matching._
import net.sf.jsqlparser.expression.Expression
import net.sf.jsqlparser.schema.Column
import play.api.libs.json.{JsString, JsValue}

import scala.language.postfixOps


final case class GroupByMatch(userArg: Option[Expression], sampleArg: Option[Expression]) extends Match {

  override type T = Expression

  override type AR = GenericAnalysisResult

  override def analyze(ua: Expression, sa: Expression): GenericAnalysisResult = GenericAnalysisResult(MatchType.SUCCESSFUL_MATCH)

  override protected def descArgForJson(arg: Expression): JsValue = JsString(arg.toString)

  override def points: Points = if (matchType == MatchType.SUCCESSFUL_MATCH) 1 halfPoint else 0 points

  override def maxPoints: Points = sampleArg match {
    case None    => 0 points
    case Some(_) => 1 halfPoint
  }

}


object GroupByMatcher extends Matcher[GroupByMatch] {

  override type T = Expression

  override protected val matchName: String = "Group Bys"

  override protected val matchSingularName: String = "des Group By-Statement"

  override protected def canMatch(e1: Expression, e2: Expression): Boolean = e1 match {
    case column1: Column => e2 match {
      case column2: Column => column1.getColumnName == column2.getColumnName
      case _               => false
    }
    case _               => false
  }


  override protected def matchInstantiation(ua: Option[Expression], sa: Option[Expression]): GroupByMatch =
    GroupByMatch(ua, sa)

}
